File: endee/reranker.py
"""
Reranking utilities for the Endee vector database client.

Each reranker accepts:
  results_map   — {field_name: [[int32_id, score], ...]}  (from SearchResult)
  field_weights — {field_name: weight}  summing to 1.0
  limit         — final number of results to return (the global search limit,
                  NOT the per-field fetch limits)
  **kwargs      — reranker-specific parameters

Returns a list of (int32_id, score) tuples sorted by score descending,
truncated to `limit`.
"""

from typing import Any, Dict, List, Optional, Tuple

from .constants import DEFAULT_RRF_RANK_CONSTANT, DEFAULT_TOPK

def _resolve_field_weights(
    field_names: List[str],
    field_weights: Optional[Dict[str, float]],
) -> Dict[str, float]:
    """Resolve and validate per-field weights against the fields present."""
    if field_weights is None:
        n = len(field_names)
        return {f: 1.0 / n for f in field_names}
    missing = [f for f in field_names if f not in field_weights]
    if missing:
        raise ValueError(f"field_weights missing entries for: {sorted(missing)}")
    total = sum(field_weights[f] for f in field_names)
    if abs(total - 1.0) > 1e-6:
        raise ValueError(f"field_weights must sum to 1.0 (got {total:.8f})")
    return field_weights


def rerank(
    search_results: Dict[str, Any],
    name: str = "rrf",
    limit: int = DEFAULT_TOPK,
    field_weights: Optional[Dict[str, float]] = None,
    rrf_k: int = DEFAULT_RRF_RANK_CONSTANT,
) -> Dict[str, Any]:
    """Fuse the per-field results from :meth:`Collection.search` into one list.

    Uses Reciprocal Rank Fusion over each field's decoded hits: a hit at rank
    ``r`` (1-based) in field ``f`` contributes ``field_weights[f] / (rrf_k + r)``
    to that object's score. Scores are summed across fields (deduped by ``id``)
    and the top ``limit`` hits are returned, sorted by fused score descending.

    Parameters
    ----------
    search_results : dict
        The return value of :meth:`Collection.search`, i.e.
        ``{"results": {field_name: [hit, ...], ...}}``.
    name : str
        Reranker algorithm. Only ``"rrf"`` is supported.
    limit : int
        Max number of fused hits to return (default ``DEFAULT_TOPK``).
    field_weights : dict, optional
        Per-field RRF weights e.g. ``{"embedding": 0.6, "keywords": 0.4}``.
        Must sum to 1.0. Defaults to uniform weighting.
    rrf_k : int
        RRF rank constant k (default 60).

    Returns
    -------
    dict  ``{"results": [hit, ...]}``  — each hit's ``similarity`` is its fused score.

    Examples
    --------
    >>> res = collection.search(fields={
    ...     "embedding": {"query": v, "limit": 50},
    ...     "keywords":  {"query": s, "limit": 50},
    ... })
    >>> fused = rerank(res, limit=10, field_weights={"embedding": 0.6, "keywords": 0.4})
    """
    if name != "rrf":
        raise ValueError("rerank: name must be 'rrf'")

    per_field = (search_results or {}).get("results")
    if not isinstance(per_field, dict):
        raise ValueError(
            "rerank: expected search results to be a per-field map of hits"
        )

    field_names = list(per_field.keys())
    if not field_names:
        raise ValueError("rerank: no fields to fuse")

    weights = _resolve_field_weights(field_names, field_weights)

    scores: Dict[str, float] = {}
    hit_by_id: Dict[str, dict] = {}
    for fname in field_names:
        weight = weights.get(fname, 0.0)
        for rank, hit in enumerate(per_field[fname] or [], start=1):
            hid = hit["id"]
            scores[hid] = scores.get(hid, 0.0) + weight / (rrf_k + rank)
            hit_by_id.setdefault(hid, hit)

    ranked = sorted(scores.items(), key=lambda x: x[1], reverse=True)[:limit]
    results = [{**hit_by_id[hid], "similarity": score} for hid, score in ranked]
    return {"results": results}

File: src/main/java/io/endee/client/Reranker.java
package io.endee.client;

import io.endee.client.types.SearchHit;
import java.util.*;

/**
 * Client-side reranking utilities for fusing per-field search results.
 *
 * <p>Example usage:
 *
 * <pre>{@code
 * Map<String, List<SearchHit>> results = collection.search(queryFields);
 * List<SearchHit> fused = Reranker.rerank(results, 10,
 *     Map.of("embedding", 0.6, "keywords", 0.4), 60);
 * }</pre>
 */
public final class Reranker {

  private static final int DEFAULT_LIMIT = 10;
  private static final int DEFAULT_RRF_K = 60;

  private Reranker() {}

  /**
   * Fuses per-field search results using Reciprocal Rank Fusion (RRF).
   *
   * @param searchResults per-field results from {@link Collection#search}
   * @param limit max number of fused hits to return
   * @param fieldWeights per-field weights (must sum to 1.0); null for uniform
   * @param rrfK RRF rank constant (default 60)
   * @return fused and sorted list of hits
   */
  public static List<SearchHit> rerank(
      Map<String, List<SearchHit>> searchResults,
      int limit,
      Map<String, Double> fieldWeights,
      int rrfK) {

    if (searchResults == null || searchResults.isEmpty()) {
      throw new IllegalArgumentException("searchResults must be a non-empty per-field map");
    }

    List<String> fieldNames = new ArrayList<>(searchResults.keySet());

    // Resolve weights
    Map<String, Double> weights;
    if (fieldWeights == null) {
      weights = new LinkedHashMap<>();
      double uniform = 1.0 / fieldNames.size();
      for (String f : fieldNames) {
        weights.put(f, uniform);
      }
    } else {
      for (String f : fieldNames) {
        if (!fieldWeights.containsKey(f)) {
          throw new IllegalArgumentException("field_weights missing entry for: " + f);
        }
      }
      double total = 0;
      for (String f : fieldNames) {
        total += fieldWeights.get(f);
      }
      if (Math.abs(total - 1.0) > 1e-6) {
        throw new IllegalArgumentException(
            "field_weights must sum to 1.0 (got " + String.format("%.8f", total) + ")");
      }
      weights = fieldWeights;
    }

    // Compute RRF scores
    Map<String, Double> scores = new LinkedHashMap<>();
    Map<String, SearchHit> hitById = new LinkedHashMap<>();

    for (String fname : fieldNames) {
      double weight = weights.getOrDefault(fname, 0.0);
      List<SearchHit> hits = searchResults.getOrDefault(fname, List.of());
      int rank = 1;
      for (SearchHit hit : hits) {
        String hid = hit.getId();
        scores.merge(hid, weight / (rrfK + rank), Double::sum);
        hitById.putIfAbsent(hid, hit);
        rank++;
      }
    }

    // Sort by score descending, take top limit
    List<Map.Entry<String, Double>> ranked = new ArrayList<>(scores.entrySet());
    ranked.sort((a, b) -> Double.compare(b.getValue(), a.getValue()));

    List<SearchHit> results = new ArrayList<>();
    for (int i = 0; i < Math.min(limit, ranked.size()); i++) {
      Map.Entry<String, Double> entry = ranked.get(i);
      SearchHit original = hitById.get(entry.getKey());
      SearchHit fused =
          new SearchHit(
              original.getId(), entry.getValue(), original.getMeta(), original.getFilter());
      results.add(fused);
    }

    return results;
  }

  /** Convenience: rerank with default limit (10) and rrfK (60). */
  public static List<SearchHit> rerank(
      Map<String, List<SearchHit>> searchResults, Map<String, Double> fieldWeights) {
    return rerank(searchResults, DEFAULT_LIMIT, fieldWeights, DEFAULT_RRF_K);
  }

  /** Convenience: rerank with uniform weights. */
  public static List<SearchHit> rerank(Map<String, List<SearchHit>> searchResults, int limit) {
    return rerank(searchResults, limit, null, DEFAULT_RRF_K);
  }
}
